app/controllers/AtsMergePageController.scala

```diff
@@ -0,0 +1,109 @@
+ */
+
+package controllers
+
+import com.google.inject.Inject
+import config.ApplicationConfig
+import connectors.DataCacheConnector
+import controllers.auth.{AuthenticatedRequest, MergePageAuthAction}
+import controllers.paye.routes.PayeAtsMainController
+import models.AtsListData
+import play.api.data.Form
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents, Result}
+import services._
+import uk.gov.hmrc.http.HttpResponse
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendController
+import uk.gov.hmrc.play.partials.FormPartialRetriever
+import uk.gov.hmrc.renderer.TemplateRenderer
+import utils.{AccountUtils, AttorneyUtils, Globals}
+import view_models.AtsForms.atsYearFormMapping
+import view_models.{AtsList, TaxYearEnd}
+import views.html.AtsMergePageView
+import views.html.errors.{GenericErrorView, TokenErrorView}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AtsMergePageController @Inject()(
+  atsMergePageService: AtsMergePageService,
+  authAction: MergePageAuthAction,
+  mcc: MessagesControllerComponents,
+  atsMergePageView: AtsMergePageView,
+  genericErrorView: GenericErrorView)(
+  implicit formPartialRetriever: FormPartialRetriever,
+  templateRenderer: TemplateRenderer,
+  appConfig: ApplicationConfig,
+  ec: ExecutionContext)
+    extends FrontendController(mcc) with AttorneyUtils with I18nSupport {
+
+  def onPageLoad: Action[AnyContent] = authAction.async { implicit request: AuthenticatedRequest[_] =>
+    if (appConfig.saShuttered && appConfig.payeShuttered)
+      Future.successful(Redirect(routes.ErrorController.serviceUnavailable().url))
+    else getSaAndPayeYearList()(request)
+
+  }
+  private def getSaAndPayeYearList(formWithErrors: Option[Form[TaxYearEnd]] = None)(
+    implicit request: AuthenticatedRequest[_]) = {
+    val session = request
+      .getQueryString(Globals.TAXS_USER_TYPE_QUERY_PARAMETER)
+      .fold(
+        request.session
+      )(parameter => request.session + (Globals.TAXS_USER_TYPE_KEY -> parameter))
+
+    atsMergePageService.getSaAndPayeYearList.map {
+      case Right(atsMergePageViewModel) =>
+        Ok(
+          atsMergePageView(
+            atsMergePageViewModel,
+            formWithErrors.getOrElse(atsYearFormMapping),
+            getActingAsAttorneyFor(
+              request,
+              atsMergePageViewModel.saData.forename,
+              atsMergePageViewModel.saData.surname,
+              atsMergePageViewModel.saData.utr)
+          ))
+          .withSession(session + ("atsList" -> atsMergePageViewModel.saData.toString))
+
+      case _ => InternalServerError(genericErrorView())
+    }
+
+  }
+
+  def onSubmit: Action[AnyContent] = authAction.async { implicit request =>
+    atsYearFormMapping.bindFromRequest.fold(
+      formWithErrors => {
+        getSaAndPayeYearList(Some(formWithErrors))(request)
+      },
+      value => {
+        val (year, atsType) = value.year.get.splitAt(4)
```

Comment:
maybe extract this behaviour out to return an AtsType

Reply:
done